File: front_end/src/services/formulaService.js
import api from '../config/api'

const formulaService = {
  async createFormulaFromMaterials(formulaData) {
    const response = await api.post('/formulas/create-from-materials', formulaData)
    return response.data.data.formula
  },

  async createVariant(formulaId, variantData) {
    const response = await api.post(`/formulas/${formulaId}/create-variant`, variantData)
    return response.data.data.formula
  },

  async getFormulas(estado = null, search = null) {
    const params = {}
    if (estado) params.estado = estado
    if (search) params.search = search

    const response = await api.get('/formulas', { params })
    return response.data.data.formulas || []
  },

  async getFormulaById(id) {
    const response = await api.get(`/formulas/${id}`)
    return response.data.data.formula
  },

  async getFormulaVersions(id) {
    const response = await api.get(`/formulas/${id}/versions`)
    return response.data.data.versions || []
  },

  async updateFormula(id, formulaData) {
    const response = await api.put(`/formulas/${id}`, formulaData)
    return response.data.data.formula
  },

  async changeEstado(id, nuevoEstado) {
    const response = await api.post(`/formulas/${id}/change-estado`, null, {
      params: { nuevoEstado }
    })
    return response.data.data.formula
  },

  async validateFormula(id) {
    const response = await api.post(`/formulas/${id}/validate`)
    return response.data.data
  },

  async deleteFormula(id) {
    const response = await api.delete(`/formulas/${id}`)
    return response.data
  }
}

export default formulaService

File: backend_java/plm/src/main/java/com/plm/plm/services/FormulaService.java
package com.plm.plm.services;

import com.plm.plm.dto.FormulaDTO;
import com.plm.plm.dto.FormulaIngredientDTO;
import com.plm.plm.dto.FormulaVersionDTO;

import java.util.List;
import java.util.Map;

public interface FormulaService {
    
    /**
     * Crear fórmula experimental desde materias primas
     */
    FormulaDTO createFormulaFromMaterials(
        String nombre,
        String objetivo,
        Double rendimiento,
        List<FormulaIngredientDTO> ingredientes,
        Integer userId
    );
    
    /**
     * Crear variante de fórmula (nueva versión)
     */
    FormulaDTO createVariant(
        Integer formulaId,
        String version,
        String justificacion,
        List<FormulaIngredientDTO> ingredientes,
        Integer userId
    );
    
    /**
     * Obtener fórmula por ID
     */
    FormulaDTO getFormulaById(Integer id);
    
    /**
     * Obtener todas las fórmulas
     */
    List<FormulaDTO> getAllFormulas(String estado, String search);
    
    /**
     * Obtener todas las versiones de una fórmula
     */
    List<FormulaVersionDTO> getFormulaVersions(Integer formulaId);
    
    /**
     * Actualizar fórmula
     */
    FormulaDTO updateFormula(Integer id, FormulaDTO formulaDTO);
    
    /**
     * Cambiar estado de fórmula
     */
    FormulaDTO changeEstado(Integer id, String nuevoEstado, Integer userId);
    
    /**
     * Eliminar fórmula
     */
    void deleteFormula(Integer id);
    
    /**
     * Validar fórmula (verificar que suma 100%, etc.)
     */
    Map<String, Object> validateFormula(FormulaDTO formula);
    
    /**
     * Obtener fórmula por ideaId
     */
    FormulaDTO getFormulaByIdeaId(Integer ideaId);

    /**
     * Crear fórmula desde una idea aprobada
     */
    FormulaDTO createFormulaFromIdea(Integer ideaId, Integer userId);
}

